diff --git a/include/linux/in.h b/include/linux/in.h
index 4065313..86a731d 100644
--- a/include/linux/in.h
+++ b/include/linux/in.h
@@ -46,6 +46,7 @@ enum {
   IPPROTO_COMP   = 108,                /* Compression Header protocol */
   IPPROTO_SCTP   = 132,		/* Stream Control Transport Protocol	*/
   IPPROTO_UDPLITE = 136,	/* UDP-Lite (RFC 3828)			*/
+  IPPROTO_HIP = 139,            /* Host Identity Protocol (RFC5201)     */
 
   IPPROTO_RAW	 = 255,		/* Raw IP packets			*/
   IPPROTO_MAX
