diff -urN linux-2.6.13.1/include/linux/net.h linux-2.6.13.1-hipmod/include/linux/net.h
--- linux-2.6.13.1/include/linux/net.h	2005-09-10 05:42:58.000000000 +0300
+++ linux-2.6.13.1-hipmod/include/linux/net.h	2006-01-31 15:37:15.000000000 +0200
@@ -26,7 +26,7 @@
 struct poll_table_struct;
 struct inode;
 
-#define NPROTO		32		/* should be enough for now..	*/
+#define NPROTO		33		/* should be enough for now..	*/
 
 #define SYS_SOCKET	1		/* sys_socket(2)		*/
 #define SYS_BIND	2		/* sys_bind(2)			*/
diff -urN linux-2.6.13.1/net/ipv6/af_inet6.c linux-2.6.13.1-hipmod/net/ipv6/af_inet6.c
--- linux-2.6.13.1/net/ipv6/af_inet6.c	2005-09-10 05:42:58.000000000 +0300
+++ linux-2.6.13.1-hipmod/net/ipv6/af_inet6.c	2006-01-31 15:34:26.000000000 +0200
@@ -864,4 +864,6 @@
 }
 module_exit(inet6_exit);
 
+EXPORT_SYMBOL(inet6_create);
+
 MODULE_ALIAS_NETPROTO(PF_INET6);
